### ISSUE TITLE
Support for .Net Core 3 Json library

### ISSUE THREAD

**Issue by monsieurleberre** [quicktype/quicktype#1347]
.Net Core 3 comes with its own set of classes to serialise and deserialise JSON, located in the `System.Text.Json` and `System.Text.Json.Serialization`.
Previous versions used to rely on `Newtonsoft.Json` and `Newtonsoft.Json.Converters` instead.

This also means some attribute names need to be adapted:
```
        [JsonProperty("id")]
        public Guid Id { get; set; }
```
becomes 
```
        [JsonPropertyName("id")]
        public Guid Id { get; set; }
```

Serialisation of enums to string can be done this way
```
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum Color { Red, Green, Blue };
```
but cannot be done by involving `JsonReader` and `JsonWriter` like it was in previous versions.

It would be nice to be able to choose between Newtonsoft.Json and System.Text.Json when using quicktype to create classes from Json.

Here is a good starting place to understand the changes I am refering to:
https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0#fast-built-in-json-support

Cheers!

**Comment by dvdsgl** [comment]
Love this idea! We would be happy to guide someone in contributing this as an option to the existing C# language support.

**Comment by Gambero81** [comment]
any news about supporting System.Text.Json?

**Comment by joelmartinez** [comment]
@dvdsgl ... I (potentially) volunteer as tribute!
![image](https://user-images.githubusercontent.com/90380/112078905-9ac4c580-8b55-11eb-873b-6d4897dfa47c.png)

Can you share any contribution guidelines/tips/insight? I'm in that brief window of time where I'm trying to decide if it would be easier to just change the generated code, or contribute this change to pay it forward :P

**Comment by dvdsgl** [comment]
@joelmartinez do it! I'll help you if you get stuck. Should not be too bad.

**Comment by Jacko1394** [comment]
Can't wait for this!! 👍

**Comment by doggy8088** [comment]
@joelmartinez Do you have any progress on this issue?

**Comment by joelmartinez** [comment]
@doggy8088 
![image](https://user-images.githubusercontent.com/90380/122069487-3bc4c480-cdc3-11eb-9238-d191e9838246.png)

Unfortunately, my "brief window of time" ended up going in another direction ... I never got quite deep enough into implementation territory here. So the task is still open for anyone else that wants a crack.

**Comment by CrAcK75** [comment]
Any news?

**Comment by bzuidgeest** [comment]
I got quicktype running with the code from the PR, seems to work fine as is for at least my needs. Is there any progress as to getting this included?

**Comment by doggy8088** [comment]
@bzuidgeest You might consider use my version of quicktype.

Source: https://github.com/doggy8088/quicktype/tree/CSharp-SystemTextJson

npm: [@willh/quicktype](https://www.npmjs.com/package/@willh/quicktype)

**Comment by doggy8088** [comment]
It because `JsonStringEnumConverter` doesn't fit my need on certain circumstance.

**Comment by bzuidgeest** [comment]
> It because `JsonStringEnumConverter` doesn't fit my need on certain circumstance.

That much was obvious. But not very helpfull. I do know it has an "issue" arround nullable enums, but personally I find enums should not be nullable, if you don't know the value a "unknown" option should be added to the enum. But opinions like that are personal and I cannot expect everyone to share them. Was this your "circumstance"?

**Comment by dylanvdmerwe** [comment]
Would be amazing to have this option built in to replace Json.NET.

**Comment by bzuidgeest** [comment]
@doggy8088 That seems like an excellent reason for doing it the way you did. The use case is rather specifc however (I think) and I don't know if it would hold up in a general use tool. What if the next person would use a different seperator and a different way of correcting it in the conversion. For example what if the replaced the dot in the enum by removing it al together? You would have to define a lot of use cases.

But thanks for explaining, that is very usefull.

**Comment by bzuidgeest** [comment]
@dylanvdmerwe Replace is a bad option. A lot of people have invested in the code as-is and rely on it. Also JSON.net is far from dead. It might fade eventually, but for now its still heavily used and has unique features that are not in System.Text.Json and might never be available there. Personally I am a proponent of what @dvdsgl suggested. The Kotlin generator has an option to choose the framework to be used. We should have that for C# to. 

I looked into the code files and we could copy the way kotlin does it and then cobine @doggy8088 repo his generator code with the existing generator code. Some name clashes would need to be fixed, but it should not be to difficult. Basically two generators and a choice between the two.

**Comment by dylanvdmerwe** [comment]
@bzuidgeest agreed my apologies I meant that having the option to select which generator to use would be incredible.

**Comment by bzuidgeest** [comment]
I took the code from the repo of @doggy8088 and merged it with the current version of quicktype using the kotlin generator as an example  of how to implement selecting between frameworks. Just like @dvdsgl sugested
It seems to work but I need to mention some caveats.
* I am NO typescript developer. But code is code so I think I did a reasonable job putting everything in place.
* The csharp testsuite still runs OK, so at least I did not break the existing stuff. But I don't know how to add a test project that can test the SystemTextJson framework to the project. Someone will have to point out what needs to happen. Could be the test docker container provided needs to be supplied with a newer .NET version, new test project etc...

code is here:
https://github.com/bzuidgeest/quicktype/tree/SystemTextJson
I could make a pull request, but I doubt that would be accepted without tests. Maybe I make one for show.

**Comment by bzuidgeest** [comment]
So I committed number of changes and as far as I can determine most tests run fine. Even some that are excluded with newtonsoft lib because of framework issues. Seems reasonable to me. Lets hope someone actually want to take a look at the PR

**Comment by Bluscream** [comment]
Why is this not a thing yet? .net 6 does not  even have newtonsoft

**Comment by bzuidgeest** [comment]
> Why is this not a thing yet? .net 6 does not even have newtonsoft

That is nonsense Newtonsoft Json.net library always has been a sperate lib and is very much supported on .NET6. Just with the nuget package. The fact that Microsoft included by default in some templates has no bearing on that. 

With the latest .net you are free to choose between the system.text.json and newtonsoft libs. 

That said, this project seems to be "on hold" as in the discussions you can read the maintainers do not have the time to maintain it.

**Comment by dvdsgl** [comment]
Fixed in #1997
